core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala

```diff
@@ -16,318 +16,212 @@
+        )
+      }
 
-        newStart -= nPadBases
+      // trim from left
+      var nLeftTrimBases = 0
+      var firstBaseFromLeftInRefAllele = allAlleles(0)(nLeftTrimBases)
+      val minAlleleLength = allAlleles.map(_.length).min
 
-        nTrimmedBasesBeforeNextPadding = 0
+      while (nLeftTrimBases < minAlleleLength - nTrimmedBasesBeforeNextPadding - 1
+        && allAlleles.forall(_(nLeftTrimBases) == firstBaseFromLeftInRefAllele)) {
 
-      } else {
+        nLeftTrimBases += 1
 
-        nTrimmedBasesBeforeNextPadding += 1
+        firstBaseFromLeftInRefAllele = allAlleles(0)(nLeftTrimBases)
+      }
 
+      allAlleles = allAlleles.map { a =>
+        a.drop(nLeftTrimBases)
+          .dropRight(nTrimmedBasesBeforeNextPadding)
       }
 
-      firstAlleleBaseFromRight = alleles(0).getBases()(
-        alleles(0).length
-        - nTrimmedBasesBeforeNextPadding - 1
-      )
-    }
+      trimSize += nTrimmedBasesBeforeNextPadding
 
-    // trim from left
-    var nLeftTrimBases = 0
-    var firstAlleleBaseFromLeft = alleles(0).getBases()(nLeftTrimBases)
-    val minAlleleLength = alleles.map(_.length).min
+      newStart += nLeftTrimBases
 
-    while (nLeftTrimBases < minAlleleLength - nTrimmedBasesBeforeNextPadding - 1
-      && alleles.forall(_.getBases()(nLeftTrimBases) == firstAlleleBaseFromLeft)) {
+      if (trimSize != 0 || nLeftTrimBases != 0) {
+        flag = true
+      }
+    }
 
-      nLeftTrimBases += 1
+    val outputRow = new GenericInternalRow(2)
 
-      firstAlleleBaseFromLeft = alleles(0).getBases()(nLeftTrimBases)
+    if (errorMessage.isEmpty) {
+      outputRow.update(
+        0,
+        InternalRow(
+          newStart,
+          end - trimSize,
+          UTF8String.fromString(allAlleles(0)),
+          ArrayData.toArrayData(allAlleles.tail.map(UTF8String.fromString(_)))
+        )
+      )
     }
 
-    alleles = alleles.map { a =>
-      Allele.create(
-        a.getBaseString()
-          .drop(nLeftTrimBases)
-          .dropRight(nTrimmedBasesBeforeNextPadding)
-          .getBytes(),
-        a.isReference
+    outputRow.update(
+      1,
+      InternalRow(
+        flag,
+        errorMessage.map(UTF8String.fromString).orNull
       )
+    )
 
-    }
+    outputRow
 
-    trimSize += nTrimmedBasesBeforeNextPadding
+  }
 
-    AlleleBlock(
-      alleles,
-      newStart + nLeftTrimBases,
-      unalignedAlleleBlock.end - trimSize
-    )
+  def isSNP(refAllele: String, altAlleles: Array[String]): Boolean = {
+    refAllele.length == 1 && altAlleles.forall(_.length == 1)
+  }
 
+  def isSymbolic(altAlleles: Array[String]): Boolean = {
+    altAlleles.exists(_.matches(".*[<|>|*].*"))
   }
 
   private val PAD_WINDOW_SIZE = 100
 
+  val normalizationResultFieldName = "normalizationResult"
+  val normalizationStatusFieldName = "normalizationStatus"
+  val changedFieldName = "changed"
+  val errorMessageFieldName = "errorMessage"
+
+  val normalizationStatusStructField = StructField(
```

Comment:
Hm, what is the upside of nesting everything in `normalizationResult` / `normalizationStatus` vs a flat struct? I generally think it's more user friendly to avoid nesting columns since it makes fields easier to access.

Reply:
I put the start, end, ref, and alts into normalizationResult as you suggested. normalizationStatus is another column. So there is only one level of nesting. 

I felt result separate from status makes more sense as it makes moving the actual result data columns, i.e., normalized (start, end, ref, alts), altogether easier after normalization. Same goes for dropping the status column after verifying no error happened or which variants changed.